File: android/app/src/main/java/com/mail/client/data/local/ThreadDao.kt
package com.mail.client.data.local

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import kotlinx.coroutines.flow.Flow

@Dao
interface ThreadDao {

    @Query("SELECT * FROM threads WHERE labelIds LIKE '%INBOX%' ORDER BY lastMessageTimestamp DESC")
    fun observeInbox(): Flow<List<ThreadEntity>>

    @Query("SELECT * FROM threads WHERE id = :id")
    suspend fun getById(id: String): ThreadEntity?

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertAll(threads: List<ThreadEntity>)

    @Delete
    suspend fun delete(thread: ThreadEntity)

    @Query("SELECT MAX(lastMessageTimestamp) FROM threads")
    suspend fun getLatestTimestamp(): Long?
}

File: flutter/lib/data/local/thread_dao.dart
part of 'mail_database.dart';

@DriftAccessor(tables: [Threads])
class ThreadDao extends DatabaseAccessor<MailDatabase> with _$ThreadDaoMixin {
  ThreadDao(super.db);

  /// Live stream of inbox threads, newest first.
  Stream<List<Thread>> watchInbox() => _watchForPattern('%INBOX%');

  /// Live stream of threads matching a label pattern (SQL LIKE), newest first.
  Stream<List<Thread>> watchForLabel(String pattern) =>
      _watchForPattern(pattern);

  Stream<List<Thread>> _watchForPattern(String pattern) =>
      (select(threads)
            ..where((t) => t.labelIds.like(pattern))
            ..orderBy([
              (t) => OrderingTerm.desc(t.lastMessageTimestamp),
            ]))
          .watch();

  Future<Thread?> getById(String id) =>
      (select(threads)..where((t) => t.id.equals(id))).getSingleOrNull();

  Future<void> upsertAll(List<ThreadsCompanion> rows) =>
      batch((b) => b.insertAllOnConflictUpdate(threads, rows));

  Future<void> deleteThread(Thread thread) => delete(threads).delete(thread);

  Future<int?> getLatestTimestamp() {
    final maxExpr = threads.lastMessageTimestamp.max();
    return (selectOnly(threads)..addColumns([maxExpr]))
        .map((row) => row.read(maxExpr))
        .getSingleOrNull();
  }

  Future<List<String>> getInboxIds() async {
    final rows = await (select(threads)
          ..where((t) => t.labelIds.like('%INBOX%'))
          ..orderBy([(t) => OrderingTerm.desc(t.lastMessageTimestamp)]))
        .get();
    return rows.map((r) => r.id).toList();
  }
}

// ── Extension helpers (mirrors ThreadEntity.labelList()) ──────────────────────

extension ThreadX on Thread {
  List<String> get labelList =>
      labelIds.isEmpty ? [] : labelIds.split(',');
}
